File: sso-demo-4/kong/plugins/oidc/handler.js
/**
 * handler.js - Kong OIDC plugin handler, JavaScript version
 *
 * Ported from the Lua nokia/kong-oidc plugin to Kong 3.x JavaScript PDK.
 * Handles OIDC authentication: introspect bearer tokens, run OIDC code flow,
 * and inject X-Userinfo / X-Access-Token / X-ID-Token headers.
 */

const utils = require("./utils");
const filter = require("./filter");
const session = require("./session");

// Mock lua-resty-openidc module - in real implementation, would call external service
// For now, we'll use Kong's native capabilities
const resty_openidc = require("resty.openidc");

function makeOidc(oidcConfig) {
  kong.log.debug(
    "OidcHandler calling authenticate, requested path: " +
      kong.request.get_path_with_query()
  );
  try {
    const res = resty_openidc.authenticate(oidcConfig);
    if (!res) {
      if (oidcConfig.recovery_page_path) {
        kong.log.debug(
          "Entering recovery page: " + oidcConfig.recovery_page_path
        );
        return kong.response.exit(302, "", {
          Location: oidcConfig.recovery_page_path,
        });
      }
      utils.exit(500, "Authentication failed");
    }
    return res;
  } catch (err) {
    if (oidcConfig.recovery_page_path) {
      kong.log.debug(
        "Entering recovery page: " + oidcConfig.recovery_page_path
      );
      return kong.response.exit(302, "", {
        Location: oidcConfig.recovery_page_path,
      });
    }
    utils.exit(500, err.message || "Authentication error");
  }
}

function introspect(oidcConfig) {
  if (!utils.hasBearerAccessToken() && oidcConfig.bearer_only !== "yes") {
    return null;
  }

  try {
    const res = resty_openidc.introspect(oidcConfig);
    if (!res) {
      if (oidcConfig.bearer_only === "yes") {
        kong.response.set_header(
          "WWW-Authenticate",
          'Bearer realm="' +
            oidcConfig.realm +
            '",error="Introspection failed"'
        );
        utils.exit(401, "Unauthorized");
      }
      return null;
    }
    kong.log.debug(
      "OidcHandler introspect succeeded, requested path: " +
        kong.request.get_path_with_query()
    );
    return res;
  } catch (err) {
    if (oidcConfig.bearer_only === "yes") {
      kong.response.set_header(
        "WWW-Authenticate",
        'Bearer realm="' + oidcConfig.realm + '",error="' + err.message + '"'
      );
      utils.exit(401, err.message || "Unauthorized");
    }
    return null;
  }
}

function handle(oidcConfig) {
  let response = null;

  if (oidcConfig.introspection_endpoint) {
    response = introspect(oidcConfig);
    if (response) {
      utils.injectUser(response);
    }
  }

  if (!response) {
    response = makeOidc(oidcConfig);
    if (response) {
      if (response.user) {
        utils.injectUser(response.user);
      }
      if (response.access_token) {
        utils.injectAccessToken(response.access_token);
      }
      if (response.id_token) {
        utils.injectIDToken(response.id_token);
      }
    }
  }
}

// Kong plugin interface
return {
  PRIORITY: 1000,
  VERSION: "1.2.4-kong3-js",

  access: function (config) {
    const oidcConfig = utils.getOptions(config);

    if (filter.shouldProcessRequest(oidcConfig)) {
      session.configure(config);
      handle(oidcConfig);
    } else {
      kong.log.debug(
        "OidcHandler ignoring request, path: " + kong.request.get_path()
      );
    }
  },
};

File: sso-demo-4/kong/plugins/oidc/handler.lua
local openidc = require("resty.openidc")
local cjson   = require("cjson")

local OidcHandler = {}
OidcHandler.PRIORITY = 1000
OidcHandler.VERSION  = "1.0.0"

local function parse_filters(csv)
  if not csv or csv == "" then return {} end
  local t = {}
  for f in csv:gmatch("[^,]+") do
    t[#t + 1] = f:match("^%s*(.-)%s*$")
  end
  return t
end

local function should_skip(filters)
  if #filters == 0 then return false end
  local path = kong.request.get_path()
  for _, pattern in ipairs(filters) do
    if path:find(pattern, 1, true) then return true end
  end
  return false
end

local function build_opts(config)
  local opts = {
    client_id              = config.client_id,
    client_secret          = config.client_secret,
    discovery              = config.discovery,
    redirect_uri_path      = config.redirect_uri_path,
    scope                  = config.scope,
    response_type          = config.response_type,
    ssl_verify             = (config.ssl_verify == "yes"),
    token_endpoint_auth_method = config.token_endpoint_auth_method,
    logout_path            = config.logout_path,
    redirect_after_logout_uri  = config.redirect_after_logout_uri,
    redirect_after_logout_with_id_token_hint = false,
    session_contents       = { id_token = true, access_token = true },
    -- Redis session storage configured programmatically (works with lua-resty-session v3 and v4)
    session_opts = {
      storage = "redis",
      redis   = { host = "redis", port = 6379, database = 0 },
      cookie  = { secure = false, http_only = true, same_site = "Lax" },
      secret  = "demo4-kong-oidc-session-secret-change-me",
    },
  }

  if config.introspection_endpoint and config.introspection_endpoint ~= "" then
    opts.introspection_endpoint = config.introspection_endpoint
    opts.introspection_endpoint_auth_method = config.introspection_endpoint_auth_method
  end
  if config.timeout then
    opts.timeout = config.timeout
  end

  -- Split discovery: browser-facing URLs stay on issuer host; Kong-internal calls
  -- (token exchange, JWKS, userinfo) use internal_host to bypass localhost routing.
  if config.internal_host and config.internal_host ~= "" then
    local issuer     = config.discovery:gsub("/%.well%-known/openid%-configuration$", "")
    local realm_path = issuer:match("^https?://[^/]*(/.+)$") or ""
    local internal   = "http://" .. config.internal_host .. realm_path

    opts.discovery = {
      issuer                       = issuer,
      authorization_endpoint       = issuer   .. "/protocol/openid-connect/auth",
      end_session_endpoint         = issuer   .. "/protocol/openid-connect/logout",
      token_endpoint               = internal .. "/protocol/openid-connect/token",
      token_introspection_endpoint = internal .. "/protocol/openid-connect/token/introspect",
      userinfo_endpoint            = internal .. "/protocol/openid-connect/userinfo",
      jwks_uri                     = internal .. "/protocol/openid-connect/certs",
      response_types_supported     = { "code" },
    }
  end

  return opts
end

function OidcHandler:access(config)
  if should_skip(parse_filters(config.filters)) then return end

  local opts = build_opts(config)
  local res, err = openidc.authenticate(opts)

  if err then
    if config.recovery_page_path and config.recovery_page_path ~= "" then
      return kong.response.exit(302, "", { Location = config.recovery_page_path })
    end
    kong.log.err("oidc: ", err)
    return kong.response.exit(500, "OIDC error: " .. tostring(err))
  end

  if res then
    if res.user then
      kong.service.request.set_header(
        "X-Userinfo",
        ngx.encode_base64(cjson.encode(res.user))
      )
    end
    if res.access_token then
      kong.service.request.set_header("X-Access-Token", res.access_token)
    end
    if res.id_token then
      kong.service.request.set_header(
        "X-ID-Token",
        ngx.encode_base64(cjson.encode(res.id_token))
      )
    end
  end
end

return OidcHandler
